File: dotnet/src/ExternalTaskWorker.cs
namespace ProcessEngine.ConsumerAPI.Client
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading;
    using System.Threading.Tasks;

    using EssentialProjects.IAM.Contracts;

    using ProcessEngine.ConsumerAPI.Contracts;
    using ProcessEngine.ConsumerAPI.Contracts.DataModel;
    using ProcessEngine.ConsumerAPI.Contracts.APIs;

    /// <summary>
    /// Periodically fetches, locks and processes ExternalTasks for a given topic.
    /// </summary>
    public class ExternalTaskWorker : IExternalTaskWorker
    {
        private const int LockDurationInMilliseconds = 30000;

        private readonly IExternalTaskConsumerApi externalTaskApi;

        /// <summary>
        /// Creates an new instance of ExternalTaskWorker
        /// </summary>
        public ExternalTaskWorker(IExternalTaskConsumerApi externalTaskApi)
        {
            this.externalTaskApi = externalTaskApi;
        }

        /// <summary>
        /// The ID of the worker
        /// </summary>
        public string WorkerId { get; } = Guid.NewGuid().ToString();

        /// <summary>
        /// Periodically fetches, locks and processes available ExternalTasks with a given topic,
        /// using the given callback as a processing function.
        /// </summary>
        /// <param name="identity">
        /// The identity to use for fetching and processing ExternalTasks.
        /// </param>
        /// <param name="topic">
        /// The topic by which to look for and process ExternalTasks.
        /// </param>
        /// <param name="maxTasks">
        /// max. ExternalTasks to fetch.
        /// </param>
        /// <param name="longpollingTimeout">
        /// Longpolling Timeout in ms.
        /// </param>
        /// <param name="handleAction">
        /// The function for processing the ExternalTasks.
        /// </param>
        public async Task WaitForHandle<TPayload>(
            IIdentity identity,
            string topic,
            int maxTasks,
            int longpollingTimeout,
            HandleExternalTaskAction<TPayload> handleAction
        )
            where TPayload : new()
        {
            const bool keepPolling = true;
            while (keepPolling)
            {
                var externalTasks = await this.FetchAndLockExternalTasks<TPayload>(
                    identity,
                    topic,
                    maxTasks,
                    longpollingTimeout
                );

                if (externalTasks.Count() == 0)
                {
                    Thread.Sleep(1000);
                    continue;
                }

                var tasks = new List<Task>();

                foreach (var externalTask in externalTasks)
                {
                    tasks.Add(this.ExecuteExternalTask<TPayload>(identity, externalTask, handleAction));
                }

                await Task.WhenAll(tasks);
            }
        }

        private async Task<IEnumerable<ExternalTask<TPayload>>> FetchAndLockExternalTasks<TPayload>(
            IIdentity identity,
            string topic,
            int maxTasks,
            int longpollingTimeout
        )
        where TPayload : new()
        {
            try
            {
                return await this.externalTaskApi.FetchAndLockExternalTasks<TPayload>(
                  identity,
                  WorkerId,
                  topic,
                  maxTasks,
                  longpollingTimeout,
                  LockDurationInMilliseconds
                );
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception);

                // Returning an empty Array here, since "waitForAndHandle" already implements a timeout, in case no tasks are available for processing.
                // No need to do that twice.
                return new List<ExternalTask<TPayload>>();
            }
        }

        private async Task ExecuteExternalTask<TPayload>(
          IIdentity identity,
          ExternalTask<TPayload> externalTask,
          HandleExternalTaskAction<TPayload> handleAction
        )
          where TPayload : new()
        {
            const int lockExtensionBuffer = 5000;
            const int lockRefreshIntervalInMs = LockDurationInMilliseconds - lockExtensionBuffer;

            var lockRefreshTimer = this.StartExtendLockTimer(identity, externalTask, lockRefreshIntervalInMs);

            try
            {
                var result = await handleAction(externalTask);

                lockRefreshTimer.Stop();

                await this.ProcessResult(identity, result, externalTask.Id);
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception);
                lockRefreshTimer.Stop();
                await this.externalTaskApi.HandleServiceError(identity, this.WorkerId, externalTask.Id, exception.Message, exception.StackTrace);
            }
        }

        private System.Timers.Timer StartExtendLockTimer<TPayload>(IIdentity identity, ExternalTask<TPayload> externalTask, int intervall)
        {
            var timer = new System.Timers.Timer(intervall);
            timer.Elapsed += async (sender, e) => await ExtendLock<TPayload>(identity, externalTask);
            timer.Start();

            return timer;
        }

        private async Task ExtendLock<TPayload>(IIdentity identity, ExternalTask<TPayload> externalTask)
        {
            try
            {
                await this.externalTaskApi.ExtendLock(identity, this.WorkerId, externalTask.Id, LockDurationInMilliseconds);
            }
            catch (Exception error)
            {
                // This can happen, if the lock-extension was performed after the task was already finished.
                // Since this isn't really an error, a warning suffices here.
                Console.WriteLine($"An error occured while trying to extend the lock for ExternalTask ${externalTask.Id}", error.Message, error.StackTrace);
            }
        }

        private async Task ProcessResult(IIdentity identity, ExternalTaskResultBase result, string externalTaskId) {

            if (result is ExternalTaskBpmnError) {

                var bpmnError = result as ExternalTaskBpmnError;
                await this.externalTaskApi.HandleBpmnError(identity, this.WorkerId, externalTaskId, bpmnError.errorCode);

            }
            else if (result is ExternalTaskServiceError<object>)
            {

                var serviceError = result as ExternalTaskServiceError<object>;
                await this
                    .externalTaskApi
                    .HandleServiceError(identity, this.WorkerId, externalTaskId, serviceError.errorMessage, serviceError.errorDetails as string);

            }
            else
            {
                await this.externalTaskApi.FinishExternalTask(identity, this.WorkerId, externalTaskId, result);
            }
        }
    }
}

File: typescript/src/external_task_worker.ts
/* eslint-disable @typescript-eslint/member-naming */
import {Logger} from 'loggerhythm';
import * as uuid from 'node-uuid';

import {HttpClient} from '@essential-projects/http';
import {IIdentity} from '@essential-projects/iam_contracts';
import {
  DataModels,
  HandleExternalTaskAction,
  IExternalTaskWorker,
} from '@process-engine/consumer_api_contracts';

import {ExternalAccessor} from './accessors/external_accessor';
import {ConsumerApiClient} from './consumer_api_client';

const logger: Logger = Logger.createLogger('processengine:consumer_api_client:external_task_worker');

export class ExternalTaskWorker<TPayload, TResultPayload> implements IExternalTaskWorker {

  private readonly _workerId = uuid.v4();
  private readonly lockDuration = 30000;
  private readonly processEngineUrl: string;
  private readonly topic: string;
  private readonly maxTasks: number;
  private readonly longpollingTimeout: number;
  private readonly processingFunction: HandleExternalTaskAction<TPayload, TResultPayload>;

  private _identity: IIdentity;
  private _pollingActive = false;
  private consumerApiClient: ConsumerApiClient;

  constructor(
    processEngineUrl: string,
    identity: IIdentity,
    topic: string,
    maxTasks: number,
    longpollingTimeout: number,
    processingFunction: HandleExternalTaskAction<TPayload, TResultPayload>,
  ) {
    this.processEngineUrl = processEngineUrl;
    this.identity = identity;
    this.topic = topic;
    this.maxTasks = maxTasks;
    this.longpollingTimeout = longpollingTimeout;
    this.processingFunction = processingFunction;

    this.initialize();
  }

  public get identity(): IIdentity {
    return this._identity;
  }

  public set identity(value: IIdentity) {
    this._identity = value;
  }

  public get workerId(): string {
    return this._workerId;
  }

  public get pollingIsActive(): boolean {
    return this._pollingActive;
  }

  public start(): void {
    this._pollingActive = true;
    this.processExternalTasks();
  }

  public stop(): void {
    this._pollingActive = false;
  }

  private initialize(): void {
    const httpClient = new HttpClient();
    httpClient.config = {url: this.processEngineUrl};

    const externalAccessor = new ExternalAccessor(httpClient);
    this.consumerApiClient = new ConsumerApiClient(externalAccessor);
  }

  private async processExternalTasks(): Promise<void> {

    while (this.pollingIsActive) {

      const externalTasks = await this.fetchAndLockExternalTasks();

      if (externalTasks.length === 0) {
        await this.sleep(1000);
        continue;
      }

      const executeTaskPromises: Array<Promise<void>> = [];

      for (const externalTask of externalTasks) {
        executeTaskPromises.push(this.executeExternalTask(externalTask));
      }

      await Promise.all(executeTaskPromises);
    }
  }

  private async fetchAndLockExternalTasks(): Promise<Array<DataModels.ExternalTask.ExternalTask<TPayload>>> {

    try {
      return this
        .consumerApiClient
        .fetchAndLockExternalTasks<TPayload>(this.identity, this.workerId, this.topic, this.maxTasks, this.longpollingTimeout, this.lockDuration);
    } catch (error) {

      logger.error(
        'An error occured during fetchAndLock!',
        error.message,
        error.stack,
      );

      // Returning an empty Array here, since "waitForAndHandle" already implements a timeout, in case no tasks are available for processing.
      // No need to do that twice.
      return [];
    }
  }

  private async executeExternalTask(externalTask: DataModels.ExternalTask.ExternalTask<TPayload>): Promise<void> {

    try {
      const lockExtensionBuffer = 5000;

      const interval = setInterval(async (): Promise<void> => this.extendLocks(externalTask), this.lockDuration - lockExtensionBuffer);

      const result = await this.processingFunction(externalTask);
      clearInterval(interval);

      await this.processResult(result, externalTask.id);

    } catch (error) {
      logger.error('Failed to execute ExternalTask!', error.message, error.stack);
      await this.consumerApiClient.handleServiceError(this.identity, this.workerId, externalTask.id, error.message, '');
    }
  }

  private async extendLocks(externalTask: DataModels.ExternalTask.ExternalTask<TPayload>): Promise<void> {
    try {
      await this.consumerApiClient.extendLock(this.identity, this.workerId, externalTask.id, this.lockDuration);
    } catch (error) {
      // This can happen, if the lock-extension was performed after the task was already finished.
      // Since this isn't really an error, a warning suffices here.
      logger.warn(`An error occured while trying to extend the lock for ExternalTask ${externalTask.id}`, error.message, error.stack);
    }
  }

  private async processResult(result: DataModels.ExternalTask.ExternalTaskResultBase, externalTaskId: string): Promise<void> {

    if (result instanceof DataModels.ExternalTask.ExternalTaskBpmnError) {
      const bpmnError = result as DataModels.ExternalTask.ExternalTaskBpmnError;
      await this.consumerApiClient.handleBpmnError(this.identity, this.workerId, externalTaskId, bpmnError.errorCode, bpmnError.errorMessage);

    } else if (result instanceof DataModels.ExternalTask.ExternalTaskServiceError) {

      const serviceError = result as DataModels.ExternalTask.ExternalTaskServiceError;
      await this.consumerApiClient.handleServiceError(
        this.identity,
        this.workerId,
        externalTaskId,
        serviceError.errorMessage,
        serviceError.errorDetails,
        serviceError.errorCode,
      );

    } else {
      await this
        .consumerApiClient
        .finishExternalTask(
          this.identity,
          this.workerId,
          externalTaskId,
          (result as DataModels.ExternalTask.ExternalTaskSuccessResult<object>).result,
        );
    }
  }

  private async sleep(milliseconds: number): Promise<void> {
    return new Promise<void>((resolve: Function): void => {
      setTimeout((): void => { resolve(); }, milliseconds);
    });
  }

}
